Mars seas5 grib download

hannahker opened the pull request:
This PR adds code to download and process archival SEAS5 data from ECMWF's MARS service. The proposed directory structure and method of calling the pipeline is **by no means set in stone** -- I'm assuming this is something we'll iterate on further as the code base grows. 

## Usage: 

The pipeline can be run locally from the command line by calling: 

```
python run_mars.py <scope> <start_year> <end_year>
```

- `<scope>`:  Either `global` or `test`. `global` will download data for the full planet and `test` will use a bounding box around Afghanistan. `test` should be used during development to download smaller subsets of data from MARS. 
- `<start_year>`: The year to begin downloading annual data for
- `<end_year>`: The year to download annual data until (not inclusive)

This code is also configured as a Job on Databricks, called "Update SEAS5 Archive". This can be triggered manually and has been used for bulk tasks (ie. more than a couple years) due to significantly improved performance. 

## Processing details

### Raw files:
Global, monthly precipitation forecasts are downloaded in yearly `.grib` files. Each raw `.grib` contains all ensemble members (26 or 51, depending on the year) and lead times (0-6 months ahead). See [this JIRA ticket](https://humanitarian.atlassian.net/browse/DSCI-539?focusedCommentId=177527) for more detailed docs on how the MARS API call is parameterized. All raw `.grib` data is stored in the `dev` Azure storage container under `global/mars/raw/`. Files are named `seas5_mars_tprate_{year}.grib`. 

### Processed files: 
The `.grib` file from each year is processed to output 84 cloud-optimized-geotiffs (`.tif`): 
1. Take the mean of all ensemble members
2. Separate by publication month and lead time
3. Set a CRS (`EPSG:4326`)

All processed files are saved to the `prod` Azure storage container under `raster/seas5/` (TODO!). Files are named `seas5_mars_tprate_em_i{pub_date}_lt{leadtime}.tif`.

NOTE: Outputs will be saved to `prod` following this PR review!

isatotun commented on constants.py:
```diff
@@ -0,0 +1,39 @@
+BBOX_GLOBAL = [-180, -90, 180, 90]
+SAS_TOKEN_DEV = os.getenv("DSCI_AZ_SAS_DEV")
+SAS_TOKEN_PROD = os.getenv("DSCI_AZ_SAS_PROD")
+CONTAINER_GLOBAL = "global"
```
I think these azure specific settings should be specified elsewhere.

hannahker replied:
Where do you think would be more appropriate?

isatotun replied:
Sorry I missed the comments following this up. I changed this on my branch, it's more along the lines of just having the one SAS_TOKEN and using that instead of the dev/prod split.

zackarno commented on run_seas5.py:
```diff
@@ -0,0 +1,27 @@
+
+    args = sys.argv
+    scope = args[1]
+    start_year = int(args[2])
+    end_year = int(args[3])
```
think we should add some additional assertions/limits on all the arguments, `scope` , `start_year`, `end_year`

here is what I've tried so far and the output which is not super clear to me:

```
python run_seas5.py test 2024 2024
#> Running SEAS5 update from 2024 to 2024 with test scope
```

```
python run_seas5.py test 2023 2023
#> Running SEAS5 update from 2023 to 2023 with test scope
```

Did runs work? 2024 shouldn't really work as it's not on MARS yet? Also what I expect should be happening is that this test run should have downloaded both sets of `.grib` files and `COG` files to a tempdir. However, how do i access/inspect that this is true?

Perhaps we should add a `dir` argument where the either both `.grib` and `.tif` or just `.tif` files can be dropped for inspection?

hannahker replied:
Great points. I've switched to using the `argparse` module, which lets us do the assertions pretty easily. I've limited the "archive" download to 2022. I've also added a `--test` flag which will download the raw and processed data locally to a `test_outputs/` directory.

zackarno reviewed (changes requested):
Hey, this is awesome!

I left some specific comments below which i think will be fairly straight forward to address.

A few general questions/comments:

- Is there `{logger}` style library (like the R one) that would be beneficial to incorporate rather than just `print()` statements?
- I think you might hit errors if you try to download year 2024 & maybe even 2023. Last time I tried 2023 it was not fully uploaded to the MARS catalogue and caused the call to error - might be fully there now since its now July & a few months later
- it's not entirely clear what is happening when I run `python run_mars.py  <scope> <start_year> <end_year>` . Is it that the `.grib` should be downloading locally to a tempdir and the COG should still get uploaded to the blob? I think for the purpose of development/testing they should probably both just get downloaded locally to a directory supplied by user for inspection - see https://github.com/OCHA-DAP/ds-raster-pipelines/pull/2#discussion_r1690228767?
- Also for the purpose of documentation - I'll note that outside of PR review, MARS outputs were also reviewed in this [quarto document](https://github.com/OCHA-DAP/ds-raster-explore/blob/main/notebooks/mars_processed_em_vs_aws.qmd) that is [rendered here](https://rpubs.com/zackarno/1204437) - and the takeaway was that the outputs looked good!

exciting stuff!

hannahker commented on seas5/download_archive_seas5_tprate.py:
```diff
@@ -0,0 +1,104 @@
+            "type": "fcmean",
+            "target": "output",
+        },
+        tp_raw,
```
Ah those commas are coming from running `black` to format the code automatically

hannahker commented on README.md:
````diff
@@ -1 +1,49 @@
+2. Install Python dependencies:
+
+```
+pip install requirements.txt
+pip install requirements-dev.txt
````
Oh nice catch!

hannahker commented on seas5/download_archive_seas5_tprate.py:
```diff
@@ -0,0 +1,104 @@
+
+
+def process_archive(tp_raw, td):
+    print(f"Processing temporary file: {tp_raw}...")
```
Yes!

hannahker commented on seas5/download_archive_seas5_tprate.py:
I'll stick with the easy option for now, which is to just add the `mars` tag

hannahker replied:
Actually will swap out `seas5` for `tprate` here, since the `seas5` is redundant with the directory name.

hannahker commented on seas5/README.md:
````diff
@@ -0,0 +1,26 @@
+The pipeline can be run locally from the command line by calling the following from the root level directory: 
+
+```
+python run_mars.py <scope> <start_year> <end_year>
````
Good catch! I'm going to keep as just `seas5` for now, since we don't have any options besides mars. If we implement a CDS download then I think we could configure that as an argument, rather than through separate files.

hannahker commented on seas5/download_archive_seas5_tprate.py:
```diff
@@ -0,0 +1,104 @@
+server = ECMWFService("mars")
+
+
+def download_archive(year, bbox, td):
```
Hmmm I'd vote for sticking with this function name since it does download the archival data for a given year. The looping just downloads archival data across multiple years. Since it's also within a more descriptive file name and directory that references `seas5`, `mars`, and `tprate`, I think adding any more specificity would be redundant.

t-downing reviewed (commented):
Great work here! Very glad we've been moving ahead with this. I've just added a few minor comments, but I think the main things I'm wondering about are:

1. It may be worth double checking that we don't want system `51`, which is what we get from CDS?
2. In general, we use `pre-commit` for our Python repos. Implementing this early on in development is much easier than bringing it in later, especially if you have alot of `.md` files. I think we should use it for this repository, since based on the name it looks like we're intending to use it for multiple raster pipelines, so keeping it well-maintained will be important. For this we need `.flake8`, `.pre-commit-config.yaml`, `pyproject.toml`, which you can just copy+paste from the repo [here](https://github.com/OCHA-DAP/ds-aa-hti-hurricanes/tree/main?tab=readme-ov-file). Then add some wording to the development section of the`README.md`. 
3. Given it looks like multiple raster pipelines will be in this repo, having a clear idea of how to organize it would be good (including the GH Actions to run the pipelines). Like, would it maybe we worth putting everything for each pipeline into a separate directory, to facilitate different requirements (and even different Python versions, if needed)? Perhaps something like this?
```
project/
├── pipelines/
│   ├── pipeline1/
│   │   ├── src/
│   │   │   ├── main.py
│   │   │   ├── datasources/
│   │   │   ├── utils/
│   │   └── requirements.txt
│   ├── pipeline2/
│   │   ├── src/
│   │   │   ├── main.py
│   │   │   ├── datasources/
│   │   │   ├── utils/
│   │   └── requirements.txt
│   └── shared/
│       ├── utils/
│       └── datasources/
├── tests/
│   ├── test_pipeline1.py
│   ├── test_pipeline2.py
├── README.md
└── .gitignore
```

hannahker commented on src/seas5/mars_tprate.py:
```diff
@@ -0,0 +1,147 @@
+            "origin": "ecmf",
+            "param": "228.172", # tprate
+            "stream": "msmm", # multi-model seasonal forecast atmospheric monthly means
+            "system": "5",
```
Through MARS I'm pretty confident it should be `5`! Check out the section on "MARS Retrievals" from [this doc](https://www.ecmwf.int/sites/default/files/medialibrary/2017-10/System5_guide.pdf). 

> All production data (re-forecasts, pre-operational and operational real-time forecasts) are
archived as ORIGIN=ECMF, SYSTEM=5, EXPVER=0001.

The request also errors out if I set to `51`.

t-downing replied:
`5` it is! I always found the `51` in CDS weird.

hannahker commented:
@t-downing @zackarno thanks both for the reviews! I've addressed most comments so should be ready for another review. 

@t-downing also totally agree that having an organized structure across pipelines will be important. However, this isn't something that I want to spend too much time designing while we're still quite early in the process of setting up these pipelines -- there's still a lot we don't know! My thinking in this PR was to keep the set up and folder structure quite straightforward and lightweight so that we can iterate as we get a better sense of requirements over time. Once this PR and @isatotun's work on IMERG are complete, I think we'll be in a much better place to plan the best way to keep things organized.

t-downing commented on src/seas5/mars_tprate.py:
```diff
@@ -0,0 +1,147 @@
+        backend_kwargs=dict(time_dims=("time", "forecastMonth")),
+    )
+
+    ds_mean = ds.mean(dim="number")
```
It seems like after taking the mean, we lose all the attributes of the `tprate` variable. We don't need most of them, but it would good to keep at least the `long_name` (Mean total precipitation rate) and `units` (m s**-1):

<img width="326" alt="image" src="https://github.com/user-attachments/assets/2aa0816b-fb02-4714-84ab-73e073fcf14e">

We can copy all the attributes over with `ds_mean["tprate"].attrs = ds["tprate"].attrs`, or just do for a subset if we don't want all the 32 existing attributes. But then at least we can immediately see the attributes when we load the COG.

hannahker replied:
Good catch! Just updated. Seems not a bad idea to just include all of them!

t-downing reviewed (commented):
Nice! Thanks for making the changes. And using `logging` is nice, good call @zackarno.

I just left one comment about preserving the attributes- at a bare minimum we want to keep the units, since sometimes with precipitation it's not immediately clear.

t-downing reviewed (approved):
Great, thanks!
